src/fides/api/middleware.py

```diff
@@ -0,0 +1,113 @@
+import json
+from typing import Any, Dict, List
+
+from fastapi import Request
+from loguru import logger
+from sqlalchemy.exc import SQLAlchemyError
+from sqlalchemy.orm import Session
+from starlette.types import Message
+
+from fides.api.api import deps
+from fides.api.ctl.sql_models import AuditLogResource  # type: ignore[attr-defined]
+from fides.api.oauth.utils import extract_client_id
+
+
+async def handle_audit_log_resource(request: Request) -> None:
+    """
+    Handles the lifecycle of recording audit log resource data.
+
+    Attempts to track the WHO, WHEN, and beginning of WHAT for
+    traceability purposes.
+
+    WHO: User ID from the API request
+    WHEN: Timestamps related to the request
+    WHAT: The endpoint, request type, and (if applicable)
+    fides_key(s) associated with the request
+    """
+
+    # details to be stored as a row on the server
+    audit_log_resource_data = {
+        "user_id": None,
+        "request_path": request.scope["path"],
+        "request_type": request.method,
+        "fides_keys": None,
+        "extra_data": None,
+    }
+    db: Session = deps.get_api_session()
+
+    # get the user id associated with the request
+    token = request.headers.get("authorization")
+    if token:
+        client = await extract_client_id(token.replace("Bearer ", ""), db)
+        audit_log_resource_data["user_id"] = client.user_id or "root"
+
+    # Access request body to check for fides_keys
+    await set_body(request, await request.body())
```

Comment:
do you mean to be calling this here and then again within `get_body()` (i.e. on L112). i could definitely be missing the point here, but this strikes me as a bit redundant - it feels like `get_body()` may have all that's needed, and this was just left in from an earlier iteration?

if i'm wrong about that, then maybe you could just update the code comments a bit to clarify specifically what's going on here, since this does feel a bit circular/redundant. or maybe i'm just being dense and not following something that's obvious 😄

Reply:
No not obvious at all! I accidentally stashed the comments instead of committing them, which linked to [this GitHub issue](https://github.com/tiangolo/fastapi/issues/394#issuecomment-883524819) that highlights this as a problem with FastAPI/Starlette and provides the workaround. Will fix this and make it clear as to why this is happening (lost half a day on this)